interfaces/prompting/requestrules/requestrules.go

```diff
@@ -1038,16 +1000,28 @@ func (rdb *RuleDB) RemoveRulesForSnapInterface(user uint32, snap string, iface s
 	return rules, nil
 }
 
-// PatchRule modifies the rule with the given ID by updating the rule's fields
-// corresponding to any of the given parameters which are set/non-empty.
+// PatchRule modifies the rule with the given ID by updating the rule's
+// constraints for any constraint field or permission which is set/non-empty.
+//
+// If the path pattern is nil, it is left unchanged from the existing rule.
+// Any permissions which are omitted from the new permissions map are left
+// unchanged from the existing rule. To remove an existing permission from
+// the rule, the permission should map to an empty permission entry.
 //
-// Any of the parameters which are equal to the default/unset value for their
-// types are left unchanged from the existing rule. Even if the given new rule
-// contents exactly match the existing rule contents, the timestamp of the rule
-// is updated to the current time. If there is any error while modifying the
-// rule, the rule is rolled back to its previous unmodified state, leaving the
-// database unchanged. If the database is changed, it is saved to disk.
-func (rdb *RuleDB) PatchRule(user uint32, id prompting.IDType, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (r *Rule, err error) {
+// Permission entries must be provided as complete units, containing both
+// outcome and lifespan (and duration, if lifespan is timespan).
+// XXX: does API unmarshalling ensures this, or do we need explicit checks?
```

Comment:
what's this XXX about exactly?

Reply:
Let's say the existing rule has the following constraints:
```json
"constraints": {
    "path-pattern": "/foo/bar",
    "permissions": {
        "read": {
            "outcome": "allow",
            "lifespan": "forever"
        },
        "write": {
            ...
        }
    }
}
```
and they want to switch the outcome of the `"read"` permission from `"allow"` to `"deny"`. We need them to include the complete permission entry for "read" in the patch constraints (though they can omit other unchanged fields, such as `"path-pattern"`, and the `"write"` permission):
```json
"constraints": {
    "permissions": {
        "read": {
            "outcome": "deny",
            "lifespan": "forever" // can't omit this
        }
    }
}
```

The resulting constraints are then:
```json
"constraints": {
    "path-pattern": "/foo/bar",
    "permissions": {
        "read": {
            "outcome": "deny",
            "lifespan": "forever"
        },
        "write": {
            ...
        }
    }
}
```

Since the only `omitempty` field in the permission entry is `"duration"`, if there's a permission entry for `"read"` in the patch constraints, the unmarshaller will enforce that it contains both the `"outcome"` and `"lifespan"` fields --- one can't omit the `"lifespan"` field and assume that the existing lifespan will be reused.

So this XXX can be removed. I'll add a comment summarizing the explanation here.